### PR TITLE
Add origin validation to popup/redirect flows

### PR DESCRIPTION
**Pull request by samhorlbeck**
(b/166287563)

### CODE REVIEW

File: packages-exp/auth-exp/src/core/util/validate_origin.ts [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+      // Do nothing if there's a URL error; just continue searching
+    }
+  }
+
+  // In the old SDK, this error also provides helpful messages.
+  fail(AuthErrorCode.INVALID_ORIGIN, { appName: auth.name });
+}
+
+function matchDomain(expected: string): boolean {
+  const currentUrl = _getCurrentUrl();
```
Comment (avolkovi) [inline packages-exp/auth-exp/src/core/util/validate_origin.ts:45]: doesn't this already do something funny on non-http? like force localhost?

File: packages-exp/auth-exp/src/core/util/validate_origin.ts [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    }
+
+    return protocol === 'chrome-extension:' && ceUrl.hostname === hostname;
+  }
+
+  if (!HTTP_REGEX.test(protocol)) {
+    return false;
+  }
+
+  if (IP_ADDRESS_REGEX.test(expected)) {
```
Comment (avolkovi) [inline packages-exp/auth-exp/src/core/util/validate_origin.ts:66]: what about port specification?

File: packages-exp/auth-exp/src/platform_browser/popup_redirect.ts [hunk 3]
```diff
@@ -19,14 +19,14 @@ import { SDK_VERSION } from '@firebase/app-exp';
 import * as externs from '@firebase/auth-types-exp';
 import { isEmpty, querystring } from '@firebase/util';
 
+import { _getProjectConfig } from '../api/project_config/get_project_config';
```
Comment (avolkovi) [inline packages-exp/auth-exp/src/platform_browser/popup_redirect.ts:22]: unused import, run your linter

File: packages-exp/auth-exp/src/core/util/validate_origin.ts (hunk 1 above)
Comment (samhorlbeck) [reply packages-exp/auth-exp/src/core/util/validate_origin.ts:45]: Nah, you're thinking of something else. This function is defined as:
```js
return (typeof self !== 'undefined' && self.location?.href) || '';
```

File: packages-exp/auth-exp/src/core/util/validate_origin.ts (hunk 2 above)
Comment (samhorlbeck) [reply packages-exp/auth-exp/src/core/util/validate_origin.ts:66]: Not sure-- this is copied directly from the old SDK fwiw

File: packages-exp/auth-exp/src/platform_browser/popup_redirect.ts (hunk 3 above)
Comment (samhorlbeck) [reply packages-exp/auth-exp/src/platform_browser/popup_redirect.ts:22]: Linter not picking it up for some reason

**Pull request merged**